src/topology/locally_constant/basic.lean

```diff
@@ -0,0 +1,285 @@
+
+initialize_simps_projections locally_constant (to_fun → apply)
+
+@[simp] lemma to_fun_eq_coe (f : locally_constant X Y) : f.to_fun = f := rfl
+
+@[simp] lemma coe_mk (f : X → Y) (h) : ⇑(⟨f, h⟩ : locally_constant X Y) = f := rfl
+
+theorem congr_fun {f g : locally_constant X Y} (h : f = g) (x : X) : f x = g x :=
+congr_arg (λ h : locally_constant X Y, h x) h
+
+theorem congr_arg (f : locally_constant X Y) {x y : X} (h : x = y) : f x = f y :=
+congr_arg (λ x : X, f x) h
+
+theorem coe_injective : function.injective (λ (f : locally_constant X Y) (x : X), f x)
+| ⟨f, hf⟩ ⟨g, hg⟩ h := have f = g, from h, by subst f
+
+@[simp, norm_cast] theorem coe_inj {f g : locally_constant X Y} : (f : X → Y) = g ↔ f = g :=
+coe_injective.eq_iff
+
+@[ext] theorem ext ⦃f g : locally_constant X Y⦄ (h : ∀ x, f x = g x) : f = g :=
+coe_injective (funext h)
+
+theorem ext_iff {f g : locally_constant X Y} : f = g ↔ ∀ x, f x = g x :=
+⟨λ h x, h ▸ rfl, λ h, ext h⟩
+
+protected lemma continuous [topological_space Y] (f : locally_constant X Y) : continuous f :=
+f.is_locally_constant.continuous
+
+/-- The constant locally constant function on `X` with value `y : Y`. -/
+def const (X : Type*) {Y : Type*} [topological_space X] (y : Y) :
+  locally_constant X Y :=
+⟨function.const X y, is_locally_constant.const _⟩
+
+lemma range_finite [compact_space X] (f : locally_constant X Y) :
+  (set.range f).finite :=
+f.is_locally_constant.range_finite
+
+lemma apply_eq_of_is_preconnected (f : locally_constant X Y) {s : set X} (hs : is_preconnected s)
+  {x y : X} (hx : x ∈ s) (hy : y ∈ s) :
+  f x = f y :=
+f.is_locally_constant.apply_eq_of_is_preconnected hs hx hy
+
+lemma apply_eq_of_preconnected_space [preconnected_space X] (f : locally_constant X Y) (x y : X) :
+  f x = f y :=
+f.is_locally_constant.apply_eq_of_is_preconnected is_preconnected_univ trivial trivial
+
+lemma eq_const [preconnected_space X] (f : locally_constant X Y) (x : X) :
+  f = const X (f x) :=
+ext $ λ y, apply_eq_of_preconnected_space f _ _
+
+lemma exists_eq_const [preconnected_space X] [nonempty Y] (f : locally_constant X Y) :
+  ∃ y, f = const X y :=
+begin
+  rcases classical.em (nonempty X) with ⟨⟨x⟩⟩|hX,
+  { exact ⟨f x, f.eq_const x⟩ },
+  { exact ⟨classical.arbitrary Y, ext $ λ x, (hX ⟨x⟩).elim⟩ }
+end
+
+/-- Push forward of locally constant maps under any map, by post-composition. -/
+def map (f : Y → Z) : locally_constant X Y → locally_constant X Z :=
+λ g, ⟨f ∘ g, λ s, by { rw set.preimage_comp, apply g.is_locally_constant }⟩
+
+@[simp] lemma map_apply (f : Y → Z) (g : locally_constant X Y) : ⇑(map f g) = f ∘ g := rfl
+
+@[simp] lemma map_id : @map X Y Y _ id = id := by { ext, refl }
+
+@[simp] lemma map_comp {Y₁ Y₂ Y₃ : Type*} (g : Y₂ → Y₃) (f : Y₁ → Y₂) :
+  @map X _ _ _ g ∘ map f = map (g ∘ f) := by { ext, refl }
+
+section comap
+
+open_locale classical
+
+variables [topological_space Y]
+
+/-- Pull back of locally constant maps under any map, by pre-composition. -/
```

Comment:
Could you please describe in the docstring what you do if `f` is not continuous (or at least mention that `(comap f g : X → Z) = g ∘ f` holds only for a continuous `f`)? BTW, the case `(hf : ¬continuous f) (H : ¬nonempty X)` never happens. I'm not sure that this is important.

Reply:
Done